ext/featurefmt/rpm/rpm.go

```diff
@@ -62,54 +118,143 @@ func (l lister) ListFeatures(files tarutil.FilesMap) ([]database.Feature, error)
 	}
 
 	// Extract binary package names because RHSA refers to binary package names.
-	out, err := exec.Command("rpm", "--dbpath", tmpDir, "-qa", "--qf", "%{NAME} %{EPOCH}:%{VERSION}-%{RELEASE}\n").CombinedOutput()
+	out, err := exec.Command("rpm", "--dbpath", tmpDir, "-qa", "--qf", "%{NAME} %{EPOCH}:%{VERSION}-%{RELEASE} %{SOURCERPM}\n").CombinedOutput()
 	if err != nil {
 		log.WithError(err).WithField("output", string(out)).Error("could not query RPM")
 		// Do not bubble up because we probably won't be able to fix it,
 		// the database must be corrupted
 		return []database.Feature{}, nil
 	}
 
+	packages := mapset.NewSet()
 	scanner := bufio.NewScanner(strings.NewReader(string(out)))
 	for scanner.Scan() {
 		line := strings.Split(scanner.Text(), " ")
-		if len(line) != 2 {
+		if len(line) != 3 {
 			// We may see warnings on some RPM versions:
 			// "warning: Generating 12 missing index(es), please wait..."
 			continue
 		}
 
-		// Ignore gpg-pubkey packages which are fake packages used to store GPG keys - they are not versionned properly.
-		if line[0] == "gpg-pubkey" {
+		if isIgnored(line[0]) {
 			continue
 		}
 
-		// Parse version
-		version := strings.Replace(line[1], "(none):", "", -1)
-		err := versionfmt.Valid(rpm.ParserName, version)
-		if err != nil {
-			log.WithError(err).WithField("version", line[1]).Warning("could not parse package version. skipping")
+		pkg := rpmPackage{PackageName: line[0]}
+		pkg.PackageVersion = strings.Replace(line[1], "(none):", "", -1)
+		if err := versionfmt.Valid(rpm.ParserName, pkg.PackageVersion); err != nil {
+			log.WithError(err).WithField("version", line[1]).Warning("skipped unparseable package")
 			continue
 		}
 
-		// Add package
-		pkg := database.Feature{
-			Name:    line[0],
-			Version: version,
+		if err := parseSourceRPM(line[2], &pkg); err != nil {
+			log.WithError(err).WithField("sourcerpm", line[2]).Warning("skipped unparseable package")
+			continue
 		}
-		packagesMap[pkg.Name+"#"+pkg.Version] = pkg
-	}
 
-	// Convert the map to a slice
-	packages := make([]database.Feature, 0, len(packagesMap))
-	for _, pkg := range packagesMap {
-		pkg.VersionFormat = rpm.ParserName
-		packages = append(packages, pkg)
+		if pkg.Valid() {
+			packages.Add(pkg)
+		}
 	}
 
-	return packages, nil
+	return convertToFeatures(packages), nil
 }
 
 func (l lister) RequiredFilenames() []string {
 	return []string{"var/lib/rpm/Packages"}
 }
+
+const (
+	terminate         = -1
+	parseRPM          = 0
+	parseArchitecture = 1
+	parseRelease      = 2
+	parseVersion      = 3
+)
+
+var stateName = map[int]string{
+	-1: "terminate",
+	0:  "RPM Token",
+	1:  "Architecture Token",
+	2:  "Release Token",
+	3:  "Version Token",
+}
```

Comment:
rather than constants with a name lookup, why not make a type for parsing states?

```go
type rpmParseState string

const (
	terminate         rpmParseState = "terminate"
	parseRPM          rpmParseState = "RPM Token"
	parseArchitecture rpmParseState = "Architecture Token"
        ...
)
```